File: why_LeetCode/Day06/Python/Next Permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return
        rfirst=len(nums)-1
        rlast=-1
        
        pivot=rfirst-1
        while pivot!=rlast and nums[pivot]>=nums[pivot+1]:
            pivot-=1
        
        if pivot==rlast:
            nums.sort()
            return
        
        for i in range(rfirst,-1,-1):
            if nums[i]>nums[pivot]:
                nums[i],nums[pivot]=nums[pivot],nums[i]
                break
        nums[pivot+1:len(nums)]=nums[pivot+1:len(nums)][::-1]
        return

File: why_LeetCode/Day06/C++/Next Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
       //官方函数
       //next_permutation(nums.begin(),nums.end());
       my_next_permutation(nums.begin(),nums.end());
    }

private:
    template<typename it>
    bool my_next_permutation(it first, it last)
    {
        //反向迭代器
        const auto rfirst=reverse_iterator<it>(last);
        const auto rlast=reverse_iterator<it>(first);
        
        //寻找满足(a<b)的ab相邻数，找到a
        auto pivot=next(rfirst);
        while(pivot!=rlast && *pivot>=*prev(pivot))
        {
            pivot++;
        }

        //如果a已是反向末尾,则说明没有下一个排列
        if(pivot==rlast)
        {
            reverse(rfirst,rlast);
            return false;
        }
        
        //寻找自右边第一个大于a的值的位置
        auto change=find_if(rfirst,rlast,bind1st(less<int>(),*pivot));
        //交换
        swap(*change,*pivot);
        //逆序
        reverse(rfirst,pivot);
        return true;
    }
};
